README.md
**Open API for Analytics** defines a set of public endpoints that enable external systems to
interact with the Distributed Data Analytics component both at the edge and at the ledger tier, and
deploy their own custom data analytics logic.

### REQUIREMENTS

* [Confluent Platform](https://www.confluent.io) >= *4.1.1*
* [MongoDB](https://www.mongodb.com/) >= *3.6.4*
* [Node.js](https://nodejs.org/) >= *10.1.0*
* [npm](https://www.npmjs.com/) >= *5.6.0*

### CLONE

    git clone git@github.com:far-edge/open-api-for-analytics.git

### CONFIGURE

Create `.env` based on `.env.example`.

    cp .env.example .env

Edit `.env`.

### CREATE THE VIRTUAL ENVIRONMENT

    cd open-api-for-analytics
    nodeenv -n 10.1.0 --prebuilt env

### ACTIVATE THE VIRTUAL ENVIRONMENT

    . env/bin/activate

### INSTALL THE DEPENDENCIES

    npm install

### RUN

    npm start

### DEACTIVATE THE VIRTUAL ENVIRONMENT

    deactivate_node

### LINT

    npm run lint

### TEST

    npm test

### GENERATE THE DOCUMENTATION

    npm run doc


CHANGELOG.md
**NOTE:** The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

**NOTE:** The project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

##### [1.0.5] - 2019-12-03

**FIXED**

* There are mistakes in the logs.
* Error occurs when a processor stops or fails.


##### [1.0.4] - 2019-12-02

**FIXED**

* Error occurs when a processor stops or fails.


##### [1.0.3] - 2019-11-12

**CHANGED**

* Exit when connection to database fails.


##### [1.0.2] - 2019-11-11

**SECURITY**

* Fix security vulnerabilities reported by GitHub.


##### [1.0.1] - 2019-09-12

**CHANGED**

* Align code with latest digital models.

**SECURITY**

* Fix security vulnerabilities reported by GitHub.


##### [1.0.0] - 2019-01-06

**ADDED**

* Create an analytics instance.
* Destroy an analytics instance.
* Update the specification of an analytics instance.
* Get an analytics instance.
* Get the state of an analytics instance.
* Get the specification of an analytics instance.
* Discover analytics instances.
* Start an analytics instance.
* Stop an anelytics instance.
* Register a data source.
* Unregister a data source.
* Discover data sources.
* Get some data from a data source.


server.js
const dblogger = require('./core/common/loggers').get('DB');
const logger = require('./core/common/loggers').get('SERVER');
const mongoose = require('mongoose');
const util = require('util');

const AIM = require('./core/workers/analytics-instance-manager');
const app = require('./app');
const blueprint = require('./blueprints/env');
const validations = require('./core/common/validations');

// Validate the environment.
const { error, value: _env } = validations.validate(process.env, blueprint);
if (error) {
  logger.error(`The environment is invalid (cause: ${ error.details[0].message }).`);
  process.exit(23);
}

// Load all models.
require('./core/models/index');

// Load all loaders.
require('./core/loaders/index');

// Load all workers.
require('./core/workers/index');

// Connect to the database.
const auth = process.env.MONGODB_USER || process.env.MONGODB_PASSWORD ?
  `${process.env.MONGODB_USER}:${process.env.MONGODB_PASSWORD}@` : '';
const host = process.env.MONGODB_HOST;
const port = process.env.MONGODB_PORT;
const name = process.env.MONGODB_NAME;
const uriwa = `mongodb://${ host }:${ port }/${ name }`;
const uri = `mongodb://${ auth }${ host }:${ port }/${ name }`;
logger.info(`System connects to the database @ ${ uriwa }.`);
mongoose.connect(uri, {
  keepAlive: 1,
  poolSize: process.env.MONGODB_POOL_SIZE,
  useNewUrlParser: true,
  useUnifiedTopology: true,
  useCreateIndex: true
}).then(() => {
  logger.info(`System connected to the database @ ${ uriwa }.`);
  // Initialise AIM.
  return AIM.init();
}).catch((error) => {
  logger.error(`System failed to connect to the database @ ${ uriwa }.`, error);
  process.exit(42);
});

// Log more details when the log level is debug.
if (process.env.LOG_LEVEL === 'debug') {
  mongoose.set('debug', (collection, method, query, doc) => {
    const iquery = util.inspect(query, false, 30);
    const idoc = util.inspect(doc, false, 30);
    dblogger.debug(`${ collection }.${ method } ${ iquery } ${ idoc }`);
  });
}

// Bind and listen for connections on the given port and host.
if (!module.parent) {
  app.listen(4444, '0.0.0.0', () => {
    logger.info(`Server started in ${process.env.NODE_ENV} mode on port 4444.`);
  });
}

module.exports = app;


blueprints/env.js
const validations = require('../core/common/validations');

const env = validations.object({
  API_BASE_URL: validations.string().uri().required(),
  LOG_LEVEL: validations.string().allow([
    'debug',
    'error',
    'info',
    'warn'
  ]).required(),
  MAX_REQUEST_BODY_SIZE: validations.number().integer().required(),
  MODEL_REPOSITORY_BASE_URL: validations.string().uri().required(),
  MONGODB_HOST: validations.string().hostname().required(),
  MONGODB_NAME: validations.string().required(),
  MONGODB_PASSWORD: validations.string().optional().empty(''),
  MONGODB_POOL_SIZE: validations.number().required(),
  MONGODB_PORT: validations.number().required(),
  MONGODB_USER: validations.string().optional().empty(''),
  NAME: validations.string().required(),
  NODE_ENV: validations.string().allow([
    'development',
    'production',
    'staging',
    'test'
  ]).required()
}).unknown().required();

module.exports = env;


core/workers/analytics-instance-manager.js
const exec = require('child-process-promise').exec;
const Promise = require('bluebird');

const AnalyticsManifest = require('../models/analytics-manifest');
const DataSourceManifest = require('../models/data-source-manifest');
const dataSources = require('../methods/data-sources');
const errors = require('../common/errors');
const { first } = require('../common/chisels');
const logger = require('../common/loggers').get('ANALYTICS-RUNTIME');
const modelDiscoverer = require('./model-discoverer');
const State = require('../models/state');

// The analytics instance table (AIT).
// The table maps analytics instance identification numbers (AIIDs) to analytics instance control
// blocks (AICBs).
const _ait = { };

// Finds a data source in the global or in any local scope.
// NOTE: We (wrongly) assume that a data source with the given ID can only exist in at most one
// scopes, which - at least in the currently implementation - is not correct.
const _findDataSource = (id) => {
  return Promise.try(() => {
    // Look for it in the global scope.
    return DataSourceManifest.findById(id);
  }).then((dataSourceManifest) => {
    if (dataSourceManifest) {
      return dataSourceManifest;
    }
    // Look for it in any local scope.
    return dataSources.discoverDataSources({ id }).then(({ dataSources }) => {
      return first(dataSources);
    });
  });
};

// Starts the analytics processor with the given manifest that is part of tha analytics instance
// with the given AIID.
const _startAnalyticsProcessor = (aiid, analyticsProcessorManifest) => {
  logger.debug(`Start analytics processor ${ analyticsProcessorManifest._id }.`);
  const dataSourceIds = analyticsProcessorManifest.dataSources.dataSource.map((ds) => {
    return ds.dataSourceManifestReferenceID;
  });
  const dataSinkId = analyticsProcessorManifest.dataSink.dataSourceManifestReferenceID;
  const analyticsProcessorDefinitionId =
    analyticsProcessorManifest.analyticsProcessorDefinitionReferenceID;
  return Promise.all([
    // Find the data source manifests.
    Promise.map(dataSourceIds, _findDataSource),
    // Find the data sink manifest.
    _findDataSource(dataSinkId),
    // Discover the analytics processor definition.
    modelDiscoverer.discoverAnalyticsProcessorDefinitions({
      id: analyticsProcessorDefinitionId
    })
  ]).spread((dataSources, dataSink, analyticsProcessorDefinitions) => {
    // Some data source does not exist.
    if (dataSources.some((ds) => { return !ds; })) {
      logger.error('One or more data sources do not exist.');
      throw new errors.BadRequestError('DATA_SOURCE_NOT_FOUND');
    }
    // The data sink does not exist.
    if (!dataSink) {
      logger.error(`Data sink ${ dataSinkId } does not exist.`);
      throw new errors.BadRequestError('DATA_SINK_NOT_FOUND');
    }
    // The analytics processor definition does not exist.
    if (!analyticsProcessorDefinitions.length) {
      const apdId = analyticsProcessorDefinitionId;
      logger.error(`Analytics processor definition ${ apdId } does not exist.`);
      throw new errors.BadRequestError('ANALYTICS_PROCESSOR_DEFINITION_NOT_FOUND');
    }
    return {
      dataSources,
      dataSink,
      analyticsProcessorDefinition: analyticsProcessorDefinitions[0]
    };
  }).then(({ dataSources, dataSink, analyticsProcessorDefinition }) => {
    // Put together the system properties.
    let properties = '';
    // Put the analytics processor ID to the system property faredge.processor.id.
    properties = `${ properties } -Dfaredge.processor.id=${ analyticsProcessorManifest._id }`;
    // Put the data sink ID to the system property faredge.sink.id.
    properties = `${ properties } -Dfaredge.sink.id=${ dataSinkId }`;
    // Put the the value for a parameter x that is passed to one of the data sources of the
    // analytics processor to the system property faredge.input.$.x, where $ is the index of the
    // data source.
    dataSources.forEach((ds, index) => {
      if (ds.dataSourceDefinitionInterfaceParameters &&
        ds.dataSourceDefinitionInterfaceParameters.parameter) {
        ds.dataSourceDefinitionInterfaceParameters.parameter.forEach((kv) => {
          properties = `${ properties } -Dfaredge.input.${ index }.${ kv.key }=${ kv.value }`;
        });
      }
    });
    // Put the value for a parameter x that is passed to the data sink of the analytics processor
    // to the system property faredge.output.x.
    if (dataSink.dataSourceDefinitionInterfaceParameters &&
      dataSink.dataSourceDefinitionInterfaceParameters.parameter) {
      dataSink.dataSourceDefinitionInterfaceParameters.parameter.forEach((kv) => {
        properties = `${ properties } -Dfaredge.output.${ kv.key }=${ kv.value }`;
      });
    }
    // Put the value for a parameter x that is passed to the analytics processor itself to the
    // system property faredge.x.
    if (analyticsProcessorManifest.parameters &&
      analyticsProcessorManifest.parameters.parameter) {
      analyticsProcessorManifest.parameters.parameter.forEach((kv) => {
        properties = `${ properties } -Dfaredge.${ kv.key }=${ kv.value }`;
      });
    }
    // Run the analytics processor.
    const command = `java -jar ${ properties } ${ analyticsProcessorDefinition.processorLocation }`;
    logger.debug(`Start analytics processor ${ analyticsProcessorManifest._id } with the command ${ command }.`);
    const p = exec(command).then(() => {
      logger.debug(`Analytics processor ${ analyticsProcessorManifest._id } stopped.`);
      // The processor stopped.
      _ait[aiid].processors[analyticsProcessorManifest._id].state = State.STOPPED;
      _ait[aiid].state = Object.values(_ait[aiid].processors).some((p) => {
        return p.state === State.FAILED;
      }) ? State.FAILED : State.STOPPED;
    }).catch((error) => {
      if (_ait[aiid].processors[analyticsProcessorManifest._id]) {
        logger.error(`Something went wrong with analytics processor ${ analyticsProcessorManifest._id }.`, error);
        _ait[aiid].processors[analyticsProcessorManifest._id].state = State.FAILED;
        _ait[aiid].state = State.FAILED;
      } else {
        logger.debug(`Analytics processor ${ analyticsProcessorManifest._id } was forced to stop.`);
      }
    });
    _ait[aiid].processors[analyticsProcessorManifest._id].process = p.childProcess;
    _ait[aiid].processors[analyticsProcessorManifest._id].state = State.RUNNING;
    logger.debug(`Started analytics processor ${ analyticsProcessorManifest._id }.`);
    return { process: p.childProcess, state: State.RUNNING };
  }).catch((error) => {
    logger.error(`Failed to start analytics processor ${ analyticsProcessorManifest._id }.`, error);
    throw error;
  });
};

// Creates the analytics instance with the given AIID.
const createAnalyticsInstance = (aiid) => {
  logger.debug(`Create analytics instance ${ aiid }.`);
  return Promise.try(() => {
    // Create the AICB for the analytics instance with the given AIID.
    const aicb = { processors: {}, state: State.STOPPED };
    // Put it into the AIT.
    _ait[aiid] = aicb;
    logger.debug(`Created analytics instance ${ aiid }.`);
    return null;
  }).catch((error) => {
    logger.error(`Failed to create analytics instance ${ aiid }.`, error);
    throw error;
  });
};

// Destroys the analytics instance with the given AIID.
const destroyAnalyticsInstance = (aiid) => {
  logger.debug(`Destroy analytics instance ${ aiid }.`);
  return Promise.try(() => {
    // Find the AICB for the analytics instance with the given AIID.
    const aicb = _ait[aiid];
    if (aicb) {
      // The analytics instance is running.
      if (aicb.state === State.RUNNING) {
        logger.error(`Analytics instance ${ aiid } is running.`);
        throw new errors.BadRequestError('ANALYTICS_INSTANCE_RUNNING');
      }
      // Remove the AICB from the AIT.
      delete _ait[aiid];
    } else {
      // There is no AICB for the given AIID.
      // Say it has been destroyed.
      logger.warn(`Analytics instance ${ aiid } has no AICB.`);
    }
    logger.debug(`Destroyed analytics instance ${ aiid }.`);
    return null;
  }).catch((error) => {
    logger.error(`Failed to destroy analytics instance ${ aiid }.`, error);
    throw error;
  });
};

// Gets the state of the analytics instance with the given AIID.
const getAnalyticsInstanceState = (aiid) => {
  logger.debug(`Gets state of analytics instance ${ aiid }.`);
  return Promise.try(() => {
    // Find the AICB for the analytics instance with the given AIID.
    const aicb = _ait[aiid];
    // There is no AICB for the given AIID.
    if (!aicb) {
      logger.error(`Analytics instance ${ aiid } has no AICB.`);
      throw new errors.NotFoundError('ANALYTICS_INSTANCE_NOT_FOUND');
    }
    logger.debug(`Got state of analytics instance ${ aiid }.`);
    return aicb.state;
  }).catch((error) => {
    logger.error(`Failed to get state of analytics instance ${ aiid }.`, error);
    throw error;
  });
};

// Initialises everything.
const init = () => {
  logger.info('Re-create all analytics instances.');
  return Promise.try(() => {
    // Find all analytics manifests.
    return AnalyticsManifest.find({});
  }).then((analyticsManifests) => {
    // Create an AICB for each one of them.
    return Promise.each(analyticsManifests, (am) => {
      logger.debug(`Re-create analytics instance ${ am._id }.`);
      _ait[am._id] = { processors: {}, state: State.STOPPED };
    });
  }).then(() => {
    logger.info('Re-created all analytics instances.');
    return null;
  });
};

// Starts the analytics instance with the given AIID.
const startAnalyticsInstance = (aiid) => {
  logger.debug(`Start analytics instance ${ aiid }.`);
  // Find the AICB for the analytics instance with the given AIID.
  const aicb = _ait[aiid];
  return Promise.try(() => {
    // Find the analytics manifest with the given ID.
    return AnalyticsManifest.findById(aiid);
  }).then((analyticsManifest) => {
    // The analytics manifest does not exist.
    if (!analyticsManifest) {
      logger.error(`Analytics instance ${ aiid } does not exist.`);
      throw new errors.NotFoundError('ANALYTICS_INSTANCE_NOT_FOUND');
    }
    // There is no AICB for the given AIID.
    if (!aicb) {
      logger.error(`Analytics instance ${ aiid } has no AICB.`);
      throw new errors.NotFoundError('ANALYTICS_INSTANCE_NOT_FOUND');
    }
    // The analytics instance is running.
    if (aicb.state === State.RUNNING) {
      logger.error(`Analytics instance ${ aiid } is running.`);
      throw new errors.BadRequestError('ANALYTICS_INSTANCE_RUNNING');
    }
    aicb.processors = analyticsManifest.analyticsProcessors.apm.reduce((acc, apm) => {
      return { ...acc, [ apm._id ]: { process: null, state: State.STOPPED } };
    }, {});
    aicb.state = State.STOPPED;
    // Start all analytics processors.
    return Promise.each(analyticsManifest.analyticsProcessors.apm, (apm) => {
      return Promise.try(() => {
        return _startAnalyticsProcessor(aiid, apm);
      }).catch((error) => {
        logger.error(`Failed to start analytics processor ${ apm._id }.`, error);
        aicb.processors[apm._id] = {
          ...aicb.processors[apm._id],
          ...{
            state: State.FAILED,
            process: null
          }
        };
        return null;
      });
    });
  }).then(() => {
    // All analytics processors have started.
    if (Object.keys(aicb.processors).every((id) => {
      return aicb.processors[id].state === State.RUNNING;
    })) {
      aicb.state = State.RUNNING;
      logger.debug(`Started analytics instance ${ aiid }.`);
      return null;
    }
    // Some of the analytics processors failed to start.
    logger.error(`Failed to start some analytics processors for analytics instance ${ aiid }`);
    // Stop the ones that started.
    return Promise.map(Object.keys(aicb.processors).filter((id) => {
      return aicb.processors[id].state === State.RUNNING;
    }), (id) => {
      aicb.processors[id].process.kill();
      return null;
    }).then(() => {
      aicb.state = Object.keys(aicb.processors).some((id) => {
        return aicb.processors[id].state === State.FAILED;
      }) ? State.FAILED : State.STOPPED;
      aicb.processors = {};
      throw new Error(`Failed to start analytics instance ${ aiid }.`);
    });
  }).catch((error) => {
    logger.error(`Failed to start analytics instance ${ aiid }.`, error);
    throw error;
  });
};

// Stops the analytics instance with the given AIID.
const stopAnalyticsInstance = (aiid) => {
  logger.debug(`Stop analytics instance ${ aiid }.`);
  // Find the AICB for the analytics instance with the given AIID.
  const aicb = _ait[aiid];
  return Promise.try(() => {
    // There is no AICB for the given AIID.
    if (!aicb) {
      logger.error(`Analytics instance ${ aiid } has no AICB.`);
      throw new errors.NotFoundError('ANALYTICS_INSTANCE_NOT_FOUND');
    }
    // The analytics instance is not running.
    if (aicb.state !== State.RUNNING) {
      logger.error(`Analytics instance ${ aiid } is not running.`);
      throw new errors.BadRequestError('ANALYTICS_INSTANCE_NOT_RUNNING');
    }
    // Stop all processors.
    // NOTE: Do your best.
    return Promise.each(Object.keys(aicb.processors), (id) => {
      return Promise.try(() => {
        aicb.processors[id].process.kill();
        return null;
      }).catch((error) => {
        logger.error(`Failed to stop analytics processor with PID ${ process.pid }.`, error);
        return null;
      });
    });
  }).then(() => {
    aicb.processors = [];
    aicb.state = State.STOPPED;
    logger.debug(`Stopped analytics instance ${ aiid }.`);
    return null;
  }).catch((error) => {
    logger.error(`Failed to stop analytics instance ${ aiid }.`, error);
    throw error;
  });
};

// Shuts everything down.
const shutDown = () => {
  logger.info('Stop all running analytics instances.');
  Object.keys(_ait).forEach((aiid) => {
    const aicb = _ait[aiid];
    if (aicb.state !== State.RUNNING) {
      return;
    }
    logger.debug(`Stop analytics instance ${ aiid }.`);
    Object.keys(aicb.processors).forEach((id) => {
      aicb.processors[id].process.kill();
    });
  });
  logger.info('Stopped all running analytics instances.');
};

module.exports = {
  createAnalyticsInstance,
  destroyAnalyticsInstance,
  getAnalyticsInstanceState,
  init,
  shutDown,
  startAnalyticsInstance,
  stopAnalyticsInstance
};


core/methods/data-sources/register-data-source.js
const Promise = require('bluebird');
const request = require('request-promise');
const rerrors = require('request-promise/errors');

const common = require('./common');
const DataSourceManifest = require('../../models/data-source-manifest');
const EdgeGateway = require('../../models/edge-gateway');
const errors = require('../../common/errors');
const logger = require('../../common/loggers').get('DATA-SOURCES');

// Registers a globally scoped data source.
const _registerGlobalDataSource = (input) => {
  logger.debug('Register a data source in the global scope.');
  return Promise.try(() => {
    // Create the data source manifest.
    const dataSourceManifest = new DataSourceManifest(input);
    // Validate the data source manifest.
    return common.validateDataSourceManifest(dataSourceManifest);
  }).then((dataSourceManifest) => {
    // Save the data source manifest.
    return dataSourceManifest.save();
  }).then((dataSourceManifest) => {
    logger.debug(`Registered data source ${ dataSourceManifest._id } in the global scope.`);
    return dataSourceManifest;
  }).catch((error) => {
    logger.error('Failed to register a data source in the global scope.', error);
    throw error;
  });
};

// Registers a locally scoped data source.
const _registerLocalDataSource = (input) => {
  logger.debug(`Register a data source in edge gateway ${ input.edgeGatewayReferenceID } local scope.`);
  return Promise.try(() => {
    // Find the edge gateway with the given ID.
    return EdgeGateway.findById(input.edgeGatewayReferenceID);
  }).then((edgeGateway) => {
    // The edge gateway does not exist.
    if (!edgeGateway) {
      logger.error(`Edge gateway ${ input.edgeGatewayReferenceID } does not exist.`);
      throw new errors.BadRequestError('EDGE_GATEWAY_NOT_FOUND');
    }
    // Forward the request to the edge gateway.
    return request({
      method: 'POST',
      uri: `${ edgeGateway.dataRouterAndPreprocessorBaseURL }/data-sources`,
      body: input,
      json: true,
      resolveWithFullResponse: true,
      simple: true
    }).catch(rerrors.StatusCodeError, (reason) => {
      if (reason.statusCode === 400) {
        throw new errors.BadRequestError(reason.response.body.error);
      }
      if (reason.statusCode === 404) {
        throw new errors.NotFoundError(reason.response.body.error);
      }
      throw new Error(reason.response.body.error);
    });
  }).then((response) => {
    // Fill in the edge gateway.
    return { ...response.body, edgeGatewayReferenceID: input.edgeGatewayReferenceID };
  }).then((dataSourceManifest) => {
    logger.debug(`Registered data source ${ dataSourceManifest._id } in edge gateway ${ input.edgeGatewayReferenceID } local scope.`);
    return dataSourceManifest;
  }).catch((error) => {
    logger.error(`Failed to register a data source in edge gateway ${ input.edgeGatewayReferenceID } local scope.`,
      error);
    throw error;
  });
};

// Registers a data source.
const registerDataSource = (input) => {
  logger.debug('Register a data source.');
  return Promise.try(() => {
    return input.edgeGatewayReferenceID ? _registerLocalDataSource(input) :
      _registerGlobalDataSource(input);
  }).then((dataSourceManifest) => {
    logger.debug(`Registered data source ${ dataSourceManifest._id }.`);
    return dataSourceManifest;
  }).catch((error) => {
    logger.error('Failed to register a data source.', error);
    throw error;
  });
};

module.exports = registerDataSource;


core/methods/data-sources/get-data-from-data-source.js
const exec = require('child-process-promise').exec;
const moment = require('moment');
const Promise = require('bluebird');

const DataSourceManifest = require('../../models/data-source-manifest');
const discoverDataSources = require('./discover-data-sources');
const errors = require('../../common/errors');
const { first } = require('../../common/chisels');
const logger = require('../../common/loggers').get('DATA-SOURCES');

// Finds a data source in the global or in any local scope.
// NOTE: We (wrongly) assume that a data source with the given ID can only exist in at most one
// scopes, which - at least in the currently implementation - is not correct.
const _findDataSource = (input) => {
  return Promise.try(() => {
    if (!input.edgeGatewayReferenceID) {
      // Look for it in the global scope.
      return DataSourceManifest.findById(input.id);
    }
    // Look for it in some local scope.
    return discoverDataSources(input).then(({ dataSources }) => {
      return first(dataSources);
    });
  });
};

// Gets data from a a data source.
const getDataFromDataSource = (input) => {
  logger.debug(`Get data from data source ${ input.id }.`);
  const N = 5;
  return Promise.try(() => {
    // Find the data source with the given ID.
    return _findDataSource(input);
  }).then((dataSource) => {
    // The data source does not exist.
    if (!dataSource) {
      logger.error(`Data source ${ input.id } does not exist.`);
      throw new errors.NotFoundError('DATA_SOURCE_NOT_FOUND');
    }
    const parameters = dataSource.dataSourceDefinitionInterfaceParameters.parameter;
    const host = first(parameters.filter((p) => { return p.key === 'host'; })).value;
    const topic = first(parameters.filter((p) => { return p.key === 'topic'; })).value;
    const port = first(parameters.filter((p) => { return p.key === 'port'; })).value;
    let command = `kafka-run-class.sh kafka.tools.GetOffsetShell --broker-list ${ host }:${ port } --topic ${ topic }`;
    return exec(command).then((result) => {
      return result.stdout.split(':')[2];
    }).then((offset) => {
      const o = isNaN(offset - N) || offset -N < 0 ? 0 : offset - N;
      command = `kafka-console-consumer.sh --bootstrap-server ${ host }:${ port } --topic ${ topic } --offset ${ o } --partition 0 --timeout-ms 1000 | grep -v ERROR | grep -v kafka.consumer.ConsumerTimeoutException`;
      return exec(command).then((result) => {
        return result.stdout.split('\n').filter((d) => { return !!d; }).map((d) => {
          const observation = JSON.parse(d);
          return { timestamp: moment(observation.collectionTimestamp).format('YYYY-MM-DD HH:mm:ss'), value: observation.value };
        });
      });
    });
  }).then((data) => {
    logger.debug(`Got data from data source ${ input.id }.`);
    return { data };
  }).catch((error) => {
    logger.error(`Failed to get data from data source ${ input.id }.`, error);
    return { data: [] };
  });
};

module.exports = getDataFromDataSource;


core/methods/analytics-instances/discover-analytics-instances.js
const Promise = require('bluebird');
const request = require('request-promise');
const rerrors = require('request-promise/errors');

const AIM = require('../../workers/analytics-instance-manager');
const AnalyticsManifest = require('../../models/analytics-manifest');
const EdgeGateway = require('../../models/edge-gateway');
const errors = require('../../common/errors');
const logger = require('../../common/loggers').get('ANALYTICS-INSTANCES');

// Discovers globally scoped analytics instances.
const _discoverGlobalAnalyticsInstances = (input) => {
  logger.debug('Discover analytics instances in the global scope.');
  return Promise.try(() => {
    if (input.edgeGatewayReferenceID) {
      return [];
    }
    // Find the analytics manifests that match the given criteria.
    return AnalyticsManifest.find({
      ...(input.id ? { _id: input.id } : { }),
      ...(input.name ? { name: input.name } : { })
    });
  }).then((analyticsManifests) => {
    // Get the state for each one of them.
    return Promise.map(analyticsManifests, (analyticsManifest) => {
      return AIM.getAnalyticsInstanceState(analyticsManifest._id).then((state) => {
        return { id: analyticsManifest._id, specification: analyticsManifest, state };
      });
    });
  }).then((analyticsInstances) => {
    logger.debug('Discovered analytics instances in the global scope.');
    return analyticsInstances;
  }).catch((error) => {
    logger.error('Failed to discover analytics instances in the global scope.', error);
    throw error;
  });
};

// Discovers locally scoped analytics instances.
const _discoverLocalAnalyticsInstances = (input) => {
  logger.debug(`Discover analytics instances in edge gateway ${ input.edgeGatewayReferenceID } local scope.`);
  return Promise.try(() => {
    // Find either all edge gateways or a specific one.
    return EdgeGateway.find({
      ...(input.edgeGatewayReferenceID ? { _id: input.edgeGatewayReferenceID } : { })
    });
  }).then((edgeGateways) => {
    // The specific edge gateway (if any) does not exist.
    if (input.edgeGatewayReferenceID && !edgeGateways.length) {
      logger.error(`Edge gateway ${ input.id } does not exist.`);
      throw new errors.BadRequestError('EDGE_GATEWAY_NOT_FOUND');
    }
    // Forward the request to all edge gateways.
    return Promise.map(edgeGateways, (edgeGateway) => {
      logger.debug(`Discover analytics instances in edge gateway ${ edgeGateway._id } local scope.`);
      return request({
        method: 'POST',
        uri: `${ edgeGateway.edgeAnalyticsEngineBaseURL }/analytics-instances/discover`,
        body: {
          ...(input.id ? { _id: input.id } : { }),
          ...(input.name ? { name: input.name } : { })
        },
        json: true,
        resolveWithFullResponse: true,
        simple: true
      }).catch(rerrors.StatusCodeError, (reason) => {
        if (reason.statusCode === 400) {
          throw new errors.BadRequestError(reason.response.body.error);
        }
        if (reason.statusCode === 404) {
          throw new errors.NotFoundError(reason.response.body.error);
        }
        throw new Error(reason.response.body.error);
      }).then((response) => {
        // Fill in the edge gateway.
        return response.body.analyticsInstances.map((analyticsInstance) => {
          return { ...analyticsInstance, edgeGatewayReferenceID: edgeGateway._id };
        });
      }).then((analyticsInstances) => {
        logger.debug(`Discovered analytics instances in edge gateway ${ edgeGateway._id } local scope.`);
        return analyticsInstances;
      }).catch((error) => {
        logger.error(`Failed to discover analytics instances in edge gateway ${ edgeGateway._id } local scope.`, error);
        return [];
      });
    });
  }).then((analyticsInstances) => {
    return analyticsInstances.reduce((acc, ams) => { return acc.concat(ams); }, []);
  }).catch((error) => {
    logger.error(`Failed to discover analytics instances in edge gateway ${ input.edgeGatewayReferenceID } local scope.`,
      error);
    throw error;
  });
};

// Discovers analytics instances.
const discoverAnalyticsInstances = (input) => {
  logger.debug('Discover analytics instances.');
  return Promise.all([
    // Discover the locally scoped analytics instances that match the given criteria.
    _discoverLocalAnalyticsInstances(input),
    // Discover the globally scoped analytics instances that match the given criteria.
    _discoverGlobalAnalyticsInstances(input)
  ]).spread((local, global) => {
    return local.concat(global);
  }).then((analyticsInstances) => {
    logger.debug('Discovered analytics instances.');
    return { analyticsInstances };
  }).catch((error) => {
    logger.error('Failed to discover analytics instances.', error);
    throw error;
  });
};

module.exports = discoverAnalyticsInstances;
